### PR TITLE
[5.7] Improve the retry jobs command

### PR DESCRIPTION
**Pull request by djug**
this PR is for improving the retry jobs command

the goal is to be able to retry jobs with more flexibility/filters

for instance, it would be possible to retry all the failed jobs that belong to a certain queue, or all the failed jobs that failed before/after a specific date/id, and it will be also possible to combine multiple filters at once

Examples:
- retry all failed jobs after a specific date:
`php artisan queue:retry --failed_at_from="2018-05-09 17:21:20"`

- retry all failed jobs before a specific date:
`php artisan queue:retry --failed_at_to="2018-05-09 17:30:00"`

- retry all failed jobs between an interval:
`php artisan queue:retry --failed_at_from="2018-05-09 17:21:20" --failed_at_to="2018-05-09 17:30:00"`
- retry all failed jobs that belongs to a specific queue:
`php artisan queue:retry --queue=jobs`

- retry all failed jobs that belongs to a specific queue and that failed after a specific date:
`php artisan queue:retry --queue=jobs  --failed_at_from="2018-05-09 17:21:20"`

- retry all failed jobs with an ID >= 40:
`php artisan queue:retry --id_from=40`

image

image

### CODE REVIEW

**Comment by sisve** [comment]
I understand that we only provide DatabaseFailedJobProvider and NullFailedJobProvider, but there may be third party packages that provides different solutions.

1. The getFiltrationOptions() needs to be added to the FailedJobProviderInterface to ensure everyone has it, to support the `$this->laravel['queue.failer']->getFiltrationOptions()`.
2. We also need to build the $signature dynamically based on getFiltrationOptions() since we do not know what the implementation actually supports.

**Comment by djug** [comment]
@sisve you are right

I added `getFiltrationOptions` to `FailedJobProviderInterface` (and to `NullFailedJobProvider`)

for the dynamic signature, I'm still thinking of a good way to achieve this that could be reusable somewhere else (either directly into the `command` class or as a trait).

File: src/Illuminate/Queue/Failed/FailedJobProviderInterface.php [hunk 1]
```diff
@@ -44,4 +44,11 @@ public function forget($id);
      * @return void
      */
     public function flush();
+
+    /**
+     * get all the options we can filter the failed jobs with
```
Comment (GrahamCampbell) [inline src/Illuminate/Queue/Failed/FailedJobProviderInterface.php]: Capital letter and full stop please. :)

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by djug** [comment]
@sisve the signature is now generated dynamically,

this feature could be used for any other command that would need to generate its signature dynamically,

**Comment by sisve** [comment]
I still see fitlration parameters declared in RetryCommand. Does this mean that those are the only one we'll support? Can a FailedJobProvider support other options that those RetryCommand supports?

**Comment by djug** [comment]
@sisve the idea was to do something like
```
 public function __construct()
    {
        parent::__construct();
        $this->dynamicParameters = $this->laravel['queue.failer']->getDynamicParameters();
    }
```
but ` $this->laravel['queue.failer']` is returning `null` when we call in from `__construct()`

any ideas about how to handle this?

**Comment by sisve** [comment]
I havn't looked into the problem, but it sounds like the constructor is executed before the QueueServiceProvider as run. What if you move the ArtisanServiceProvider (part of ConsoleSupportServiceProvider) to somewhere after the QueueServiceProvider in your config/app.php?

Other solutions may include building a custom InputDefinition that queries the queue.failer for the information. I'm not sure when in the execution process the definition is called and if the queue.failer is set at that time.

**Comment by djug** [comment]
@sisve changing the order in `config/app.php` didn't help much

**Comment by taylorotwell** [comment]
Even if I agreed with everything in this PR, I would not be able to merge it because it does not follow formatting conventions, methods are missing doc blocks, etc.

**Comment by djug** [comment]
@taylorotwell would you merge it if I fix the formatting issues, and add the doc blocks?

**Comment by kyranb** [comment]
Any chance on this being added in?

**Pull request closed without merging**